PR: https://github.com/odoo/odoo/pull/

From: 7e008469e4e5afe9b4c7151a4738240462359f98
From: Yannick Tivisse
Date: 2017-05-26 11:13:12

Breaking data model changes scores: del:2 + add:0, change matches:
-    _inherit = 'ir.ui.menu'
-    load_xmlid = fields.Boolean(default=False)

Total Changes: 39

[IMP] base: Always load menuitem xmlids on templates

PURPOSE
=======

Currently we can choose to load the xmlid of a menuitem if web_tour is installed by setting the attribute `load_xmlid=True` on the view.

This is strange for 2 reasons:
* Searching all the ir.model.data on menuitems with load_xmlid=True to compute its xmlid, searching the related subtree on the returned structure of the method load_menus is quite complicated for the perfomance gain (if there is any perfomance gain to do it)
* This is not working in community in frontend, as only the root menuitems are loaded, thus it is the `load_menus_root` instead of the `load_menus` that is called. This issue prevent a tour step to be displayed correctly on community and enterprise on a menuitem.

SPECIFICATION
=============

Remove all the `load_xmlid` attributes and the related mechanism and move it into base to always load the xmlids for all the menuitems. It won't bee to heavy as the result is cached.

It's better the keep the current mechanism (Load the xmlids from ir.model.data and insert them into the tree) as we only make 1 search to the ir.model.data

================================= pseudo patch: =================================

--- a/addons/web_tour/models/__init__.py
+++ b/addons/web_tour/models/__init__.py
@@ -2,5 +2,4 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import ir_http
-from . import ir_ui_menu
 from . import tour

--- a/addons/web_tour/models/ir_ui_menu.py
+++ b/None
@@ -1,38 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, fields, models, tools
-from odoo.tools import pycompat
-
-
-class IrUiMenu(models.Model):
-    _inherit = 'ir.ui.menu'
-
-    load_xmlid = fields.Boolean(default=False)
-
-    @api.model
-    @tools.ormcache_context('self._uid', 'debug', keys=('lang',))
-    def load_menus(self, debug):
-        """ Extends load_menus to include requested xmlids """
-        menu_root = super(IrUiMenu, self).load_menus(debug)
-
-        menu_ids = [menu.id for menu in self.browse(menu_root['all_menu_ids']) if menu.load_xmlid]
-        xmlids = {
-            d.res_id: d.module + "." + d.name
-            for d in self.env['ir.model.data'].sudo().search([('res_id', 'in', menu_ids), ('model', '=', 'ir.ui.menu')])
-        }
-
-        def _find_subtree(tree, node_id):
-            # Returns the subtree whose id is node_id
-            if tree['id'] == node_id:
-                return tree
-            else:
-                for child in tree['children']:
-                    subtree = _find_subtree(child, node_id)
-                    if subtree:
-                        return subtree
-
-        for menu_id, menu_xmlid in pycompat.items(xmlids):
-            _find_subtree(menu_root, menu_id)['xmlid'] = menu_xmlid
-
-        return menu_root
